<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Creates, Merges, and Reads Binary Dosage Files • BinaryDosage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Creates, Merges, and Reads Binary Dosage Files">
<meta property="og:description" content="Tools to create binary dosage from either vcf
    or gen files, merge binary dosage files, and read
    binary dosage files.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">BinaryDosage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/how-to-use.html">Vignette Title</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="binarydosage-convert-imputed-snp-data-to-a-binary-dosage-format" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#binarydosage-convert-imputed-snp-data-to-a-binary-dosage-format" class="anchor"></a>BinaryDosage: Convert Imputed SNP data to a Binary Dosage Format</h1></div>
</div>
<div id="bdose-format" class="section level1">
<h1 class="hasAnchor">
<a href="#bdose-format" class="anchor"></a>BDose Format</h1>
<div id="introduction" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h3>
<p>Genotype imputation is an essential tool in genomics, enabling association testing with markers not directly genotyped, increasing statistical power, and facilitating data pooling between studies that employ different genotyping platforms. Two commonly used software packages for imputation are <a href="https://genome.sph.umich.edu/wiki/Minimac">minimac</a> and <a href="http://mathgen.stats.ox.ac.uk/impute/impute_v2.html">Impute2</a>. Furthermore, services such as the <a href="https://imputationserver.sph.umich.edu/index.html">Michigan Imputation Server</a> have made genotype imputation much more accessible and streamlined.</p>
<p>While a number of software options are available for analyses of imputed data (e.g. PLINK, EPACTS), fewer are available for Genomewide Gene x Environment Interaction Scan (GWIS). Furthermore, data management tasks such as parsing, subsetting, and merging, while manageable in smaller studies, quickly become unyieldy and prohibitively slow with very large samples sizes. We aim to address these limitations by converting imputation outputs into a binary data format, BDose (Binary Dosage). The benefits of a binary format are two fold - decreased hard drive storage requirements (compared to a VCF file), and speed of parsing/analyses. The BinaryDosage package contains functions to convert VCF and Impute2 formatted files into BDose format, along with functions to merge and subset samples/SNPs.</p>
<p>For GWAS/GWIS analysis of BDose files, please refer to the <a href="https://github.com/USCbiostats/GxEScanR"><strong>GxEScanR</strong></a> package.</p>
</div>
<div id="description" class="section level3">
<h3 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h3>
<div id="bdose-files-contain-at-minimum-the-following-information" class="section level5">
<h5 class="hasAnchor">
<a href="#bdose-files-contain-at-minimum-the-following-information" class="anchor"></a>BDose files contain at minimum the following information:</h5>
<ul>
<li>Subject IDs</li>
<li>SNP information</li>
<li>Chromosome number</li>
<li>Location in base pairs</li>
<li>Reference Allele</li>
<li>Alternate Allele</li>
<li>Dosage values</li>
</ul>
</div>
<div id="bdose-files-may-contain-any-of-the-following-information" class="section level5">
<h5 class="hasAnchor">
<a href="#bdose-files-may-contain-any-of-the-following-information" class="anchor"></a>BDose files may contain any of the following information:</h5>
<ul>
<li>Family IDs</li>
<li>SNP information</li>
<li>SNP name (rs number)</li>
<li>Minor allele indicator (reference of alternate allele)</li>
<li>Minor allele frequency</li>
<li>Average call rate</li>
<li>r2</li>
<li>Genetic probabilities, Pr(g=0), Pr(g=1), Pr(g=2)</li>
</ul>
</div>
<div id="merged-bdose-files-contain-the-following-information" class="section level5">
<h5 class="hasAnchor">
<a href="#merged-bdose-files-contain-the-following-information" class="anchor"></a>MERGED BDose files contain the following information:</h5>
<ul>
<li>Number of files combined</li>
<li>Number of subjects in each VCF file</li>
</ul>
</div>
</div>
<div id="functions" class="section level3">
<h3 class="hasAnchor">
<a href="#functions" class="anchor"></a>Functions</h3>
<ul>
<li>
<strong>VCFtoBD</strong> - Convert VCF/Impute2 files to BDose format</li>
<li>
<strong>MergeBD</strong> - Merge BDose files</li>
<li>(Currently retains only intersection of markers based on <em>Chromosome:BasePair_RefAllele_AltAllele</em>)</li>
<li>
<strong>GetBDoseInfo</strong> - Create R List containing BDose file attributes (required for <strong>GetSNPValues</strong>)</li>
<li>
<strong>GetSNPValues</strong> - Obtain genotype Dosages/Genotype Probabilities, outputs results to an R Data Frame</li>
</ul>
</div>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<ol>
<li>Install the <a href="https://github.com/hadley/devtools">devtools</a> package</li>
<li>Install the <a href="https://github.com/USCbiostats/BinaryDosage">BinaryDosage</a> package directly from the USCbiostats repository on GitHub:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(devtools)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"USCbiostats/BinaryDosage"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BinaryDosage)</a></code></pre></div>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<div id="general-workflow" class="section level4">
<h4 class="hasAnchor">
<a href="#general-workflow" class="anchor"></a>General Workflow</h4>
<div id="binarydosage" class="section level5">
<h5 class="hasAnchor">
<a href="#binarydosage" class="anchor"></a><a href="https://github.com/USCbiostats/BinaryDosage"><strong>BinaryDosage</strong></a>
</h5>
<ul>
<li>Convert VCF to BDose</li>
</ul>
</div>
<div id="gxescanr" class="section level5">
<h5 class="hasAnchor">
<a href="#gxescanr" class="anchor"></a><a href="https://github.com/USCbiostats/GxEScanR"><strong>GxEScanR</strong></a>
</h5>
<ul>
<li>Use <em>ReadBDInfo()</em> to create an R list of BDose file attributes</li>
<li>Note: This is the only way to see what is in a BDose file</li>
<li>
<em>read.table</em> phenotype data (<em>R</em>)</li>
<li>Note: order of covariates matter, refer to package documentation</li>
<li>Run GxEScanR to conduct GWAS/GWIS</li>
</ul>
</div>
</div>
<div id="example" class="section level4">
<h4 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h4>
<p>Example datasets <em>batch1.vcf.gz</em> and <em>batch2.vcf.gz</em> are representative of VCF output files obtained from the Michigan Imputation Server, and contain the following:</p>
<table class="table">
<thead><tr class="header">
<th>Study</th>
<th>N</th>
<th>Number of SNPs</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>B1</td>
<td>100</td>
<td>500</td>
</tr>
<tr class="even">
<td>B2</td>
<td>100</td>
<td>500</td>
</tr>
</tbody>
</table>
<div id="convert-vcf-to-bdose-format" class="section level6">
<h6>Convert VCF to BDose Format</h6>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BinaryDosage)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># Currently, must extract compressed files (vcf.gz support </span><span class="al">TBD</span><span class="co">)</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co"># system('gzcat batch1.vcf.gz &gt; batch1.vcf') system('gzcat batch2.vcf.gz &gt;</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># batch2.vcf')</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co"># Convert VCF Batch 1 and Batch 2 files:</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="kw">VCFtoBD</span>(<span class="st">"batch1.vcf"</span>, <span class="st">"batch1.bdose"</span>)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="kw">VCFtoBD</span>(<span class="st">"batch2.vcf"</span>, <span class="st">"batch2.bdose"</span>)</a></code></pre></div>
<!-- ``` {r, eval = T, message = F, collapse = T} -->
<!-- bdose1 <- GetBinaryDosageInfo("./example/batch1.bdose") -->
<!-- -->
<!-- bdose1$NumSamples -->
<!-- bdose1$NumSNPs -->
<!-- ``` -->
</div>
<div id="merge-bdose-files" class="section level6">
<h6>Merge BDose files</h6>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BinaryDosage)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">filesToMerge &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"batch1.bdose"</span>, <span class="st">"batch2.bdose"</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw">MergeBD</span>(<span class="st">"merged.bdose"</span>, filesToMerge)</a></code></pre></div>
</div>
<div id="verify-contents" class="section level6">
<h6>Verify contents</h6>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">pooled &lt;-<span class="st"> </span><span class="kw">GetBDoseInfo</span>(<span class="st">"merged.bdose"</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">pooled<span class="op">$</span>NumSamples</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">#&gt; [1] 200</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">pooled<span class="op">$</span>NumSNPs</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; [1] 500</span></a></code></pre></div>
</div>
<div id="gwasgwis---see-gxescanr" class="section level6">
<h6>GWAS/GWIS - See <a href="https://github.com/USCbiostats/GxEScanR"><strong>GxEScanR</strong></a>
</h6>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>John Morrison <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by John Morrison.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
